framework/src/main/java/org/checkerframework/framework/util/DefaultQualifierKindHierarchy.java

```diff
@@ -0,0 +1,802 @@
+import org.plumelib.util.UtilPlume;
+
+/**
+ * This is the default implementation of {@link QualifierKindHierarchy}.
+ *
+ * <p>By default, the subtyping information and information about polymorphic qualifiers is read
+ * from meta-annotations on the annotation classes. This information is used to infer further
+ * information such as top and bottom qualifiers. Subclasses can override the following methods to
+ * change this behavior:
+ *
+ * <ul>
+ *   <li>{@link #createQualifierKinds(Collection)}
+ *   <li>{@link #createDirectSuperMap()}
+ *   <li>{@link #initializePolymorphicQualifiers()}
+ *   <li>{@link #initializeQualifierKindFields(Map)}
+ *   <li>{@link #createLubsMap()}
+ *   <li>{@link #createGlbsMap()}
+ * </ul>
+ *
+ * {@link DefaultQualifierKindHierarchy.DefaultQualifierKind} is the implementation used for {@link
+ * QualifierKind} by this class.
+ */
+@AnnotatedFor("nullness")
+public class DefaultQualifierKindHierarchy implements QualifierKindHierarchy {
+
+    /**
+     * A mapping from canonical name of a qualifier class to the QualifierKind object representing
+     * that class.
+     */
+    protected final Map<@Interned String, DefaultQualifierKind> nameToQualifierKind;
+
+    /** A list of all {@link QualifierKind}s sorted in ascending order. */
+    protected final List<DefaultQualifierKind> qualifierKinds;
+
+    /** All the qualifier kinds that are the top qualifier in their hierarchy. */
+    private final Set<DefaultQualifierKind> tops;
+
+    /** All the qualifier kinds that are the bottom qualifier in their hierarchy. */
+    private final Set<DefaultQualifierKind> bottoms;
+
+    /**
+     * Holds the lub of qualifier kinds. {@code lubs.get(kind1).get(kind2)} returns the lub of kind1
+     * and kind2.
+     */
+    private final Map<QualifierKind, Map<QualifierKind, QualifierKind>> lubs;
+
+    /**
+     * Holds the glb of qualifier kinds. {@code lubs.get(kind1).get(kind2)} returns the glb of kind1
+     * and kind2.
+     */
+    private final Map<QualifierKind, Map<QualifierKind, QualifierKind>> glbs;
+
+    @Override
+    public Set<? extends QualifierKind> getTops() {
+        return tops;
+    }
+
+    @Override
+    public Set<? extends QualifierKind> getBottoms() {
+        return bottoms;
+    }
+
+    @Override
+    @SuppressWarnings("nullness:dereference.of.nullable") // All QualifierKinds are keys in lubs.
+    public @Nullable QualifierKind leastUpperBound(QualifierKind q1, QualifierKind q2) {
+        return lubs.get(q1).get(q2);
+    }
+
+    @Override
+    @SuppressWarnings("nullness:dereference.of.nullable") // All QualifierKinds are keys in glbs.
+    public @Nullable QualifierKind greatestLowerBound(QualifierKind q1, QualifierKind q2) {
+        return glbs.get(q1).get(q2);
+    }
+
+    @Override
+    public List<? extends QualifierKind> allQualifierKinds() {
+        return qualifierKinds;
+    }
+
+    @Override
+    public @Nullable QualifierKind getQualifierKind(String name) {
```

Comment:
No, this method could be used to check if an annotation is a qualifier in this hiearachy.

Reply:
My comment was a bit unclear, sorry.  I was asking whether it is ever used that way in the current code.  If not, maybe it could throw the exception to simplify all current clients.  (That would require creating a different method if we ever come up with a client that needs to do just a check.)